### ISSUE TITLE
HttpRecorder leaking file handles

### ISSUE THREAD

**Issue by anjackson** [ukwa/heritrix3#2]
In our domain crawl, the HttpRecorder is leaking a significant number of file handles. It seems to roughly correlate with the number of DNS successes (i.e. it's almost as large as that), and manifests as multiple open file handles to the same file, e.g.

```
[root@crawler03 ~]# ls -l /proc/4727/fd/ | grep tt169http | head
lr-x------ 1 heritrix heritrix 64 Sep 18 16:02 10033 -> /heritrix/scratch/sixmonthly/tt169http.ris
lr-x------ 1 heritrix heritrix 64 Sep 18 16:02 100341 -> /heritrix/scratch/sixmonthly/tt169http.ris
lr-x------ 1 heritrix heritrix 64 Sep 18 16:02 100359 -> /heritrix/scratch/sixmonthly/tt169http.ris
lr-x------ 1 heritrix heritrix 64 Sep 18 16:02 100584 -> /heritrix/scratch/sixmonthly/tt169http.ris
lr-x------ 1 heritrix heritrix 64 Sep 18 16:02 10090 -> /heritrix/scratch/sixmonthly/tt169http.ris
lr-x------ 1 heritrix heritrix 64 Sep 18 16:02 101032 -> /heritrix/scratch/sixmonthly/tt169http.ris
lr-x------ 1 heritrix heritrix 64 Sep 18 16:02 101051 -> /heritrix/scratch/sixmonthly/tt169http.ris
...
```

(or better, via `lsof -p PID`)

The name narrows it down to the RecordingInputStream used in the ToeThread HttpRecorder.

It's not clear exactly what's happening, but I propose:
1. To add a finaliser to the HttpRecorder class to close the file handles when it gets GCd.
2. To attempt to find the path that is failing to close the recorder and remedy that.
3. To add hooks that check if the RIS file has been closed before it is overwritten with a new one, and close it first.

**Comment by anjackson** [comment]
Ah, may be something to do with our ViralContentProcessor, which re-uses the file.

**Comment by anjackson** [comment]
Yes, the ViralContentProcessor was not closing the InputStream that it was passing to the ClamAV daemon. Generally, this didn't matter because the items are cached in RAM only, but for large items they get cached to disk and it is that minority of resources that cause a file handle to be re-allocated on each virus scan.

**Comment by anjackson** [comment]
Resolved in ukwa/bl-heritrix-modules@3df62af79f7878e620169b06e350ca840dc139f5
